src/main/java/com/xiaomi/infra/pegasus/client/PegasusTable.java

```diff
@@ -985,6 +1036,94 @@ public void batchGet(List<Pair<byte[], byte[]>> keys, List<byte[]> values, int t
+      throws PException {
+    if (keys == null || keys.size() == 0) {
+      throw new PException("Invalid parameter: keys should not be null or empty");
+    }
+    if (results == null) {
+      throw new PException("Invalid parameter: results should not be null");
+    }
+    results.clear();
+    List<Future<BatchGetResult>> futures = new ArrayList<Future<BatchGetResult>>();
+
+    // Group By Partition Index Id
+    List<batch_get_request> requestList = new ArrayList<batch_get_request>();
+    List<PException> partitionToPException = new ArrayList<>();
+    for (int i = 0; i < this.table.getPartitionCount(); i++) {
+      batch_get_request templateRequest = new batch_get_request();
+      templateRequest.keys = new ArrayList<>();
+      requestList.add(templateRequest);
+      partitionToPException.add(null);
+    }
+
+    List<Integer> responseIndex = new ArrayList<>();
+    for (Pair<byte[], byte[]> k : keys) {
+      byte[] hashKey = k.getLeft();
+      byte[] sortKey = k.getRight();
+
+      long partitionHash = table.getKeyHash(hashKey);
+      gpid gpid = table.getGpidByHash(partitionHash);
+      batch_get_request r = requestList.get(gpid.get_pidx());
+      responseIndex.add(gpid.get_pidx());
+
+      full_key fullKey = new full_key();
+      fullKey.hash_key = new blob(hashKey);
+      fullKey.sort_key = new blob(sortKey);
+      r.keys.add(fullKey);
+    }
+
+    List<Boolean> emptyResults = new ArrayList<>();
+    for (int i = 0; i < requestList.size(); i++) {
+      batch_get_request request = requestList.get(i);
+      if (request.keys.isEmpty()) {
+        emptyResults.add(true);
+        futures.add(null);
+        continue;
+      }
+      emptyResults.add(false);
+      futures.add(asyncBatchGet(request, timeout));
+    }
+
+    List<Map<Pair<String, String>, byte[]>> resultMapList = new ArrayList<>();
+    for (int i = 0; i < this.table.getPartitionCount(); i++) {
+      if (emptyResults.get(i)) {
+        Map<Pair<String, String>, byte[]> emptyMap = new HashMap<>();
+        resultMapList.add(emptyMap);
+        continue;
+      }
+      Future<BatchGetResult> fu = futures.get(i);
+      fu.awaitUninterruptibly();
+      if (fu.isSuccess()) {
+        resultMapList.add(fu.getNow().valueMap);
+      } else {
+        Throwable cause = fu.cause();
+        partitionToPException.set(
+            i, new PException("Get value of keys[" + i + "] failed: " + cause.getMessage(), cause));
+      }
+    }
+
+    int count = 0;
+    for (int i = 0; i < responseIndex.size(); i++) {
+      int index = responseIndex.get(i);
+      if (null != partitionToPException.get(index)) {
+        results.add(Pair.of(partitionToPException.get(index), null));
+        continue;
+      }
+
+      byte[] value =
+          resultMapList
+              .get(index)
+              .get(Pair.of(new String(keys.get(i).getLeft()), new String(keys.get(i).getRight())));
```

Comment:
`Pair.of` will return new object, though it's value is same. Do you make sure `map` can get value by the key?

Reply:
why not the map cannot get the value by the key ? is there any pitfall?  my  unit test pass, it looks ok;

Reply:
OK, I see,  `Pair` overload `hashCode()` and `equal` method